File: Algorithm_101_Symmetric Tree.java
// 思路：直接递归比较两节点的值是否相等以及子节点的对称情况
// 注意null的处理

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return judge(root.left, root.right);
    }
    public boolean judge(TreeNode n1, TreeNode n2) {
        if (n1 == null && n2 == null) {
            return true;
        } else if (n1 == null || n2 == null) {
            return false;
        }
        return n1.val == n2.val && judge(n1.right, n2.left) && judge(n1.left, n2.right);
    }
}

File: Algorithm_101_Symmetric Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.helper(root.left, root.right)
    
    def helper(self, left: TreeNode, right: TreeNode) -> bool:
        if not left and not right:
            return True
        elif not left and right or not right and left:
            return False
        return left.val == right.val and \
            self.helper(left.left, right.right) and \
            self.helper(left.right, right.left)